docs: Add back end E2E test document (#1381)
apache/apisix-dashboard: docs/back-end-e2e.md

# See the License for the specific language governing permissions and
# limitations under the License.
#
-->

This document describes how to use E2E test locally.

## Start with source code

1. To run back end E2E test, please start the `manager-api`, `admin-api`, etcd and upstream node at first.

2. To start the `manager-api` project locally, please refer to [develop](./develop.md) web section.

3. To start the etcd locally, please refer to [etcd start](https://github.com/apache/apisix/blob/master/doc/install-dependencies.md) web section.

4. To start the `admin-api` project locally, please refer to [admin-api start](https://github.com/apache/apisix#get-started) web section.

5. To start the upstream node locally, please install docker in the local environment and execute the command.

   ```sh
    docker run -d --name upstream -v /(Your apisix-dashboard folder path)/api/test/docker/upstream.conf:/etc/nginx/conf.d/default.conf:ro -p 80:80 -p 1980:1980 -p 1981:1981 -p 1982:1982 -p 1983:1983 -p 1984:1984 --ip 172.16.238.20 johz/upstream:v2.0
   ```

6. After all the services are started, you can start the back-end E2E test.

> 7. The upstream node IP is temporarily changed to the local IP address. After the test, it should be changed to GitHub upstream node IP. If the test case does not involve the upstream node, it does not need to be modified.

Reviewer: @Jaycean Since within this step, developers are required to change every single occurrence for IP address of upstream nodes. Cause I believe that we need to cover all test cases even for only few changes. So I'm afraid this is not a rather practical way. How about just guide developers to setup docker-compose environment for local e2e testing? Thanks.

Reviewer 2: I use docker-compose to run local e2e testing. when I update the main code, just delete the image of Manager API and let it rebuild. building the image of Manager API only takes a little time.

Reviewer: > I use docker-compose to run local e2e testing. when I update the main code, just delete the image of Manager API and let it rebuild. building the image of Manager API only takes a little time.

@nic-chen yep, from my perspective, yours is much better and even easier. Thanks for sharing.

Author: I updated the steps of using docker compose for E2E testing. It's really unfriendly to start the method locally according to the source code, so is this method left for developers to choose, or is it just run with docker compose?

Reviewer 3: Good. Using docker-compose for E2E is a better way.

Reviewer: @Jaycean current doc is awesome. Thanks for the update.

# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
-->

This document describes how to use E2E test locally.

## Start with source code

1. To run back end E2E test, please start the `manager-api`, `admin-api`, etcd and upstream node at first.

2. To start the `manager-api` project locally, please refer to [develop](./develop.md) web section.

3. To start the etcd locally, please refer to [etcd start](https://github.com/apache/apisix/blob/master/doc/install-dependencies.md) web section.

> 4. To start the `admin-api` project locally, please refer to [admin-api start](https://github.com/apache/apisix#get-started) web section.

Author: done

<!--
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
-->

> This document describes how to use E2E test locally.

Reviewer 4: ```suggestion
This document describes how to run E2E test locally.
```

Author: done

> 1. The E2E test using the back end needs to start the manage-API, admin-API, etcd and upstream node.

Reviewer 4: ```suggestion
1. To run back end E2E test, please start the `manage-api`, `admin-api`, etcd and upstream node at first.
```

Author: done

> 2. To start the manage-API project locally, please refer to [develop](./develop.md) web section.

Reviewer 4: ```suggestion
2. To start the `manager-api` project locally, please refer to [develop](./develop.md) web section.
```

Author: done

3. To start the etcd locally, please refer to [etcd start](https://github.com/apache/apisix/blob/master/doc/install-dependencies.md) web section.

> 4. To start the admin-API project locally, please refer to [admin-API start](https://github.com/apache/apisix#get-started) web section.

Reviewer 4: ```suggestion
4. To start the `admin-api` project locally, please refer to [admin-api start](https://github.com/apache/apisix#get-started) web section.
```

Author: done

5. To start the upstream node locally, please install docker in the local environment and execute the command.

   ```sh
    docker run -d --name upstream -v /(Your apisix-dashboard folder path)/api/test/docker/upstream.conf:/etc/nginx/conf.d/default.conf:ro -p 80:80 -p 1980:1980 -p 1981:1981 -p 1982:1982 -p 1983:1983 -p 1984:1984 johz/upstream:v2.0
   ```

6. After all the services are started, you can start the back-end E2E test.

7. The upstream node IP is temporarily changed to the local IP address. After the test, it should be changed to GitHub upstream node IP. If the test case does not involve the upstream node, it does not need to be modified.

   ```sh
    # Local E2E test create route example
    {
        "uris": ["/test-test"],
        "name": "route_all",
        "desc": "所有2",
        "methods": ["GET"],
        "hosts": ["test.com"],
        "status": 1,
        "upstream": {
            "nodes": {
                # upstream node IP is required for local test
                "(local ip):1981": 1
                        },
            "type": "roundrobin"

>             }

Reviewer 4: it would be better to format the code

Author: done

    }

     # GitHub E2E test create route example
    {
        "uris": ["/test-test"],
        "name": "route_all",

>         "desc": "所有2",

Reviewer: @Jaycean we would better use English words here. Thanks.

Author: done，thks

        "methods": ["GET"],
        "hosts": ["test.com"],
        "status": 1,
        "upstream": {
            "nodes": {
                "172.16.238.20:1981": 1
            },
            "type": "roundrobin"
         }
    }
   ```

8. Enter the E2E folder and execute the command to test all E2E test files.

   ```sh
    cd /(Your apisix-dashboard folder path)/api/test/e2e
    go test -v
   ```

9. You can also do E2E test on a single file.

   ```sh
    cd /(Your apisix-dashboard folder path)/api/test/e2e
    go test -v E2E-test-file.go base.go

>    ```

Reviewer 3: Hello, IMO, sometimes we need the delete the etcd store info, otherwise, it will make the test failed.
I think this can also be added to the doc.

Author: Done. Very good advice. Thks. I added a note to remind developers.
